apps/calendar/js/backend/store/alarm.js

```diff
@@ -0,0 +1,285 @@
+  /**
+   * Finds single alarm by busytime id.
+   *
+   * @param {Object} related busytime object.
+   * @param {IDBTransaction} [trans] optional transaction.
+   * @param {Function} callback node style [err, records].
+   */
+  findAllByBusytimeId: function(busytimeId, trans, callback) {
+    if (typeof(trans) === 'function') {
+      callback = trans;
+      trans = null;
+    }
+
+    if (!trans) {
+      trans = this.db.transaction(this._dependentStores);
+    }
+
+    var store = trans.objectStore(this._store);
+    var index = store.index('busytimeId');
+    var key = IDBKeyRange.only(busytimeId);
+
+    index.mozGetAll(key).onsuccess = function(e) {
+      callback(null, e.target.result);
+    };
+  },
+
+  /**
+   * Works queue putting alarms into the alarm api database where needed.
+   *
+   */
+  workQueue: function(now, callback) {
+    if (typeof(now) === 'function') {
+      callback = now;
+      now = null;
+    }
+
+    now = now || new Date();
+    var alarms = navigator.mozAlarms;
+
+    if (!alarms) {
+      if (callback) {
+        callback(null);
+      }
+
+      return;
+    }
+
+    var self = this;
+    var requiresAlarm = false;
+
+    /**
+     * Why are we getting all alarms here?
+     *
+     * The alarms are designed to keep the total number
+     * of entires (in mozAlarms) down but we should keep at
+     * minimum one active at all times. For example if the user
+     * has sync turned off and wants notifications we need
+     * to have an alarm go off to trigger adding more alarms.
+     */
+    var req = alarms.getAll();
+
+    //XXX: even with the good reasons above we need
+    //     to justify the perf cost here later.
+    req.onsuccess = function(e) {
+      var data = e.target.result;
+      var len = data.length;
+      var mozAlarm;
+
+      requiresAlarm = true;
+
+      for (var i = 0; i < len; i++) {
+        mozAlarm = data[i].data;
+        if (
+          mozAlarm &&
+          'eventId' in mozAlarm &&
+          'trigger' in mozAlarm
+        ) {
+          requiresAlarm = false;
+          break;
+        }
+      }
+
+      callback = callback || function() {};
+      self._moveAlarms(
+        now,
+        requiresAlarm,
+        callback
+      );
+    };
+
+    req.onerror = function() {
+      var msg = 'failed to get alarms';
+      console.error('CALENDAR:', msg);
+
+      if (callback) {
+        callback(new Error(msg));
+      }
+    };
+  }
+};
+
+function dispatchAlarms(past, future) {
+  // If the alarm was meant to be triggered in the past,
+  // we want to immediately issue a notification.
+  // However, in bug 857284 we add the stipulation that
+  // we shouldn't issue duplicates, so handle that here also.
+  var eventToAlarm = {};
+  past.forEach(alarm => {
+    var event = alarm.eventId;
+    if (!event || event in eventToAlarm) {
+      return;
+    }
+
+    eventToAlarm[event] = alarm;
+  });
+
+  /* TODO(gareth): Should not talk to notifications controller from backend.
```

Comment:
Want to call out that we'll have to change these to events / stream of some sort eventually